Changed the translate behavior to support translation engines.

burzum opened the pull request:
This is the base to implement other translation engines like [shadow table translations](https://github.com/AD7six/cakephp-shadow-translate) that perform better than the existing key / value lookup when you have a lot of data.

Commit:
Refactoring the TranslateBehavior to use engines.

Commit:
Adding an abstract TranslationEngine base class for translations.

Commit:
phpcs fixes.

Commit:
Implemented a helper method in the TranslateBehaviorTest.

The reason for that is the preparation for being able to extend the test by another test and simply change the defaults for the behavior and run it against another translation engine.

markstory commented:
Are we planning on having multiple engines in core? If not I don't know if this is really necessary.

markstory commented on src/ORM/Behavior/Translate/EavEngine.php:
```diff
@@ -0,0 +1,431 @@
+            $field,
+            $locale,
+            $query,
+            $select
```
Indentation is funny here.

burzum replied:
I'm pretty sure its the code from the original behavior, how would you like to have it look like?

burzum commented:
@markstory if we don't want to add the shadow table engine to the core please let me know *now* before I waste time on it. The key / value store based solution that comes with the core isn't always the best option.

Commit:
phpcs fix.

Commit:
Fixes and cleanup of the default config values for the TranslateBehavior.

The config is now mostly in the translation engines, so keeping some of the config fields in the behavior overwrote always the defaults of another engine.

Commit:
Removing a field from a test case that doesn't exist in the used fixture.

TranslateBehaviorTest::testConditions() used a "description" field that doesn't even exist in the used Articles fixture.

ADmad commented:
I am :+1: on having shadow tables based translation in core, never liked the eav based implementation.

AD7six commented:
This looks pretty good to me - I could be vetoed, but whether it's merged before 3.0 or goes in 3.1, I would say it's not going to be wasted effort @burzum.

The same treatment could/should be applied to the tree behavior - which I should have done a long, long time ago. But didn't :blush:.

burzum commented:
Thanks for the feedback guys, then lets vote on this. I'll open another PR for the shadow table *after* a decision was made if this goes into the core or not. If not I'll make a PR for @AD7six plugin or give him the code.

netstyler commented:
:+1: as I need to use shadow table behavior in cake 2.x because of so much translation data. I could migrate quiet easy to 3.x with this in the core... ;)

markstory commented:
I'm on board with having the shadow table implementation, I just wanted to make sure the intention was to include shadow table in core. Having an engine with only one implementation feels wrong. But if we are going to have multiples then it totally makes sense.

Given that 3.0.0 is literally around the corner (3-4ish weeks) I'm thinking it might be best to move this into 3.1 vs. 3.0. Merging in large possibly disruptive changes during an RC process is something I'd prefer to avoid.

lorenzo commented:
I'm ok with adding the shadow tables engine for 3.1

davidyell commented:
I like the idea of this shadow table translation feature. Sounds like a much easier and more elegant way to do translations.

So I'm,
 :+1: On merge into core
:+1: On merge for 3.1 release

Commit:
Adding tests for the ShadowTableEngine and the engine itself.

Commit:
Removing not used "uses" from the TranslateBehaviorTest.

lorenzo commented:
Please reopen this targeting 3.1

burzum commented:
Merging this with the 3.1 branch is going to become pretty hard. Just tried it. :(

ionas commented:
3.1 Roadmap says: 
> Engine based TranslateBehavior support EAV and shadow tables #5958

https://github.com/cakephp/cakephp/wiki/3.1-Roadmap

Did this happen?
Is there support for shadow tables now (and cakefest just ignored that?)

ADmad commented:
@ionas No it did not happen. Shadow tables based translation is supported through plugin [ad7six/shadow-translate](https://github.com/AD7six/cakephp-shadow-translate)

ionas commented:
I don't know how to edit wiki pages on GH but this should then change

> 3.1 will be backwards compatible with 3.0.
> 
>     √ ORM Improvements
>         √ notMatching() For exclusionary filtering using associations. #6102
>         √ leftJoinWith() and innerJoinWith() to join associations for filtering, without loading associated data. #6101
>         ?? Lazy eager loading of associations. #6104 (jeremy harris?)
>     √ Route CLI tool for debugging and inspecting application routes. #4810
>     √ Multiple messages for FlashComponent #6161
>     √ Shell Helpers - Add Table and Progress bar helpers.
>     ?? Exception stack traces in JSON responses #6000
>     ~~Engine based TranslateBehavior support EAV and shadow tables #5958~~
>     ?? Update exception rendering to use a stack of handlers. This will allow more flexible error response generation for applications with API's and allow internal separation of development and non-development errors.
>     √ Deprecate Session helper
>     ?? Add Auth helper (requires AuthComponent refactoring into request object)
On https://github.com/cakephp/cakephp/wiki/3.1-Roadmap

I did not check the 3.0-Roadmap for ages but the 3.2 and 3.3 Roadmaps seem fine.
